File: 0363-Max Sum of Rectangle No Larger Than K/0363-Max Sum of Rectangle No Larger Than K.cpp
class Solution {
public:
    int maxSumSubmatrix(vector<vector<int>>& matrix, int k) {
        int m = matrix.size();
        int n = matrix[0].size();
        
        int maxSum = INT_MIN;
        for (int l = 0; l < n; ++l)
        {
            vector<int> sums(m);
            vector<int> presums(1 + m);
            vector<int> temp(1 + m);
            for (int r = l; r < n; ++r)
            {
                for (int i = 0; i < m; ++i)
                {
                    sums[i] += matrix[i][r];
                    presums[i + 1] = presums[i] + sums[i];
                }
                
                mergeSort(presums, 0, m, temp, k, maxSum);
                if (maxSum == k)
                {
                    return k;
                }
            }
        }
        
        return maxSum;
    }
    
private:
    void mergeSort(vector<int>& presums, int start, int end, vector<int>& temp, int k, int& maxSum)
    {
        if (start < end)
        {
            int mid = start + (end - start) / 2;
            mergeSort(presums, start, mid, temp, k, maxSum);
            mergeSort(presums, mid + 1, end, temp, k, maxSum);
            
            int i = start, j = mid + 1;
            int q = start;
            int p = mid + 1;
            while (i <= mid)
            {
                while (p <= end && presums[p] - presums[i] <= k)
                {
                    maxSum = max(maxSum, presums[p] - presums[i]);
                    ++p;
                }
                
                while (j <= end && presums[j] < presums[i])
                {
                    temp[q++] = presums[j++];
                }
                
                temp[q++] = presums[i++];
            }
            
            while (j <= end)
            {
                temp[q++] = presums[j++];
            }
            
            for (int l = start; l <= end; ++l)
            {
                presums[l] = temp[l];
            }
        }
    }
};

File: 0301-0400/0363-Max Sum of Rectangle No Larger Than K/0363-Max Sum of Rectangle No Larger Than K.py
import bisect
import math
class Solution:
    def maxSumSubmatrix(self, matrix: 'List[List[int]]', k: 'int') -> 'int':
        m = len(matrix)
        n = len(matrix[0])
        maxSum = -math.inf

        for l in range(n):
            sums = [0] * m
            for r in range(l, n):
                table = [0]
                presum = 0
                for i in range(m):
                    sums[i] += matrix[i][r]
                    presum += sums[i]
                    index = bisect.bisect_left(table, presum - k)
                    if index != len(table):
                        maxSum = max(maxSum, presum - table[index])
                        if maxSum == k:
                            return k
                    bisect.insort_left(table, presum)                

        return maxSum


class Solution2:
    def maxSumSubmatrix(self, matrix: 'List[List[int]]', k: 'int') -> 'int':
        m = len(matrix)
        n = len(matrix[0])
        maxSum = -math.inf
        presum = [[0] * (n + 1) for _ in range(m)]
        for i in range(m):
            for j in range(n):
                presum[i][j + 1] = presum[i][j] + matrix[i][j]


        for l in range(n):
            for r in range(l, n):
                curr = 0
                table = [0]
                for i in range(m):
                    curr += presum[i][r + 1] - presum[i][l]
                    index = bisect.bisect_left(table, curr - k)
                    if index != len(table):
                        maxSum = max(maxSum, curr - table[index])
                        if maxSum == k:
                            return k
                    bisect.insort_left(table, curr)                

        return maxSum
